draft-ietf-tls-esni.md

```diff
@@ -663,52 +662,61 @@ may actually be invalid for one or the other ClientHello, in which case a fresh
 ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
 Otherwise, the usual rules for HelloRetryRequest processing apply.
 
-Clients bind encryption of the second ClientHelloInner to encryption of the
-first ClientHelloInner via the derived ech_hrr_key by running a modified
-encryption process:
+The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
+the second ClientHelloOuter for any referenced extensions. It then encrypts
+the new EncodedClientHelloInner value as a second message with the previous
+HPKE context:
 
 ~~~
-    pkR = Deserialize(ECHConfig.public_key)
-    enc, context = SetupPSKS(pkR, "tls ech" || 0x00 || ECHConfig,
-                             ech_hrr_key, "hrr key")
     payload = context.Seal(ClientHelloOuterAAD,
                            EncodedClientHelloInner)
 ~~~
 
-The `info` parameter to SetupPSKS is the concatenation of "tls ech", a
-zero byte, and the serialized ECHConfig. ClientHelloOuterAAD is computed from
-the second ClientHelloOuter as described in {{authenticating-outer}}. Note the
-encrypted value is also authenticated by ech_hrr_key. The rationale for this
-is described in {{flow-hrr-hijack}}.
+ClientHelloOuterAAD is computed as described in {{authenticating-outer}}, but
+again using the second ClientHelloOuter. Note an HPKE context maintains a
+sequence number, so this operation internally uses a fresh nonce for the AEAD
+operation. Reusing the HPKE context avoids an attack described in
+{{flow-hrr-hijack}}.
+
+The client then modifies the "encrypted_client_hello" extension in
+ClientHelloOuter as follows:
+
+- `cipher_suite` is unchanged and contains the client's chosen HPKE cipher
+  suite.
+- `config_id` is replaced with the empty string.
+- `enc` is replaced with the empty string.
+- `payload` is replaced with the value computed above.
 
 If the client offered ECH in the first ClientHello, then it MUST offer ECH in
 the second. Likewise, if the client did not offer ECH in the first ClientHello,
 then it MUST NOT not offer ECH in the second.
 
-[[OPEN ISSUE: Should we be using the PSK input or the info input?  On the one
-hand, the requirements on info seem weaker, but maybe actually this needs to be
-secret? Analysis needed.]]
-
 ## GREASE ECH {#grease-ech}
 
 If the client attempts to connect to a server and does not have an ECHConfig
 structure available for the server, it SHOULD send a GREASE {{?RFC8701}}
-"encrypted_client_hello" extension as follows:
+"encrypted_client_hello" extension in the first ClientHello as follows:
 
-- Set the "cipher_suite" field to a supported ECHCipherSuite. The selection
+- Set the `cipher_suite` field to a supported ECHCipherSuite. The selection
```

Comment:
I have no idea when we use quotation marks or backticks, but we seem to use backticks when referring to ClientECH fields?

Reply:
I've been going with backticks for structures and their fields and quotes for extension names.